src/hints/state.rs

```diff
@@ -0,0 +1,372 @@
+    vm.insert_value(bottom_addr, node_values[2])?;
+
+    // TODO: prevent underflow (original hint doesn't appear to care)?
+    // compute `ids.hash_ptr.result = ids.node - ids.edge.length`
+    let res = node - node_values[0];
+
+    // ids.hash_ptr refers to SpongeHashBuiltin (see cairo-lang's sponge_as_hash.cairo)
+    // it is a struct with 6 felts and `result` is the 4th
+    let hash_ptr = get_relocatable_from_var_name(vars::ids::HASH_PTR, vm, ids_data, ap_tracking)?;
+    let hash_ptr_result: Relocatable = (hash_ptr + 3)?;
+    vm.insert_value(hash_ptr_result, res)?;
+
+    // TODO: __patricia_skip_validation_runner
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use rstest::{fixture, rstest};
+
+    use super::*;
+    use crate::io::input::CommitmentInfo;
+
+    #[fixture]
+    fn os_input() -> StarknetOsInput {
+        StarknetOsInput {
+            contract_state_commitment_info: CommitmentInfo {
+                previous_root: 1_usize.into(),
+                updated_root: 2_usize.into(),
+                tree_height: 251_usize.into(),
+                commitment_facts: Default::default(),
+            },
+            contract_class_commitment_info: CommitmentInfo {
+                previous_root: 11_usize.into(),
+                updated_root: 12_usize.into(),
+                tree_height: 251_usize.into(),
+                commitment_facts: Default::default(),
+            },
+            deprecated_compiled_classes: Default::default(),
+            compiled_classes: Default::default(),
+            contracts: Default::default(),
+            class_hash_to_compiled_class_hash: Default::default(),
+            general_config: Default::default(),
+            transactions: Default::default(),
+            block_hash: Default::default(),
+            compiled_class_visited_pcs: Default::default(),
+        }
+    }
+
+    #[rstest]
+    fn test_set_preimage_for_state_commitments(os_input: StarknetOsInput) {
+        let mut vm = VirtualMachine::new(false);
+        vm.add_memory_segment();
+        vm.add_memory_segment();
+        vm.add_memory_segment();
+        vm.set_fp(3);
+
+        let ap_tracking = ApTracking::new();
+        let constants = HashMap::new();
+
+        let ids_data = HashMap::from([
+            (vars::ids::INITIAL_ROOT.to_string(), HintReference::new_simple(-3)),
+            (vars::ids::FINAL_ROOT.to_string(), HintReference::new_simple(-2)),
+            (vars::ids::MERKLE_HEIGHT.to_string(), HintReference::new_simple(-1)),
+        ]);
+        insert_value_from_var_name(vars::ids::MERKLE_HEIGHT, 251_usize, &mut vm, &ids_data, &ap_tracking)
+            .expect("Couldn't insert 252 into ids.MERKLE_HEIGHT");
+
+        let mut exec_scopes: ExecutionScopes = Default::default();
+        exec_scopes.insert_value(vars::scopes::OS_INPUT, os_input);
+
+        set_preimage_for_state_commitments(&mut vm, &mut exec_scopes, &ids_data, &ap_tracking, &constants).unwrap();
```

Comment:
Shouldn't we call a hint through hint processor to verify it is properly registered there?

Reply:
This is a unit test so I don't really care about that here. If a hint is not registered integration tests will not work and the error is easy to spot and resolve, so I'd rather keep the unit tests focused.